### PR TITLE
Return to main menu if shaders fail to compile, and display an error message.

### LINKED ISSUE #10097: Do not join the game with broken shaders
##### Minetest version

```
Minetest 5.3.0-dev-d7a211389 (GNU/Linux)
```

##### OS / Hardware

Operating system: lubuntu 18.04
CPU: intel

GPU model: intel
OpenGL version: core profile 4.5, version string 3.0

##### Summary

If the shaders are broken, Minetest shows everything in dark and gives an error message.
I think nobody wants to play game with broken shaders, so it could be better to abort joining and go back to the main menu if shaders are broken.
In the settings menu it is then possible to disable the shaders.

##### Steps to reproduce

* Change the nodes fragment shader so that it cannot be successfully compiled 
* Join a world in Minetest

### PR DESCRIPTION
**Pull request by lisacvuk**
Fixes #10097 

## To do
- [x] Find a way to display an error for object shaders as well.  

This PR is ready for review.

## How to test
Add some random text to any shader, open a world.

### CODE REVIEW

File: src/nodedef.cpp [hunk 1]
```diff
@@ -1375,12 +1381,19 @@ void NodeDefManager::updateTextures(IGameDef *gamedef,
 
 	u32 size = m_content_features.size();
 
+	bool success = true;
+
 	for (u32 i = 0; i < size; i++) {
 		ContentFeatures *f = &(m_content_features[i]);
-		f->updateTextures(tsrc, shdsrc, meshmanip, client, tsettings);
+		if (!f->updateTextures(tsrc, shdsrc, meshmanip, client, tsettings))
+			success = false;
```
Comment (hecktest) [inline src/nodedef.cpp:1389]: Does this loop need to complete in case of failure?

Comment (lisacvuk) [reply src/nodedef.cpp]: It doesn't - my bad. Thanks for pointing it out :) Fixed

File: src/client/client.cpp [hunk 2]
```diff
@@ -1768,7 +1768,14 @@ void Client::afterContentReceived()
 	tu_args.last_percent = 0;
 	tu_args.text_base =  wgettext("Initializing nodes");
 	tu_args.tsrc = m_tsrc;
-	m_nodedef->updateTextures(this, texture_update_progress, &tu_args);
+
+	if (!m_nodedef->updateTextures(this, texture_update_progress, &tu_args)) {
+		error_message = "Failed to update textures! Your shader code is probably wrong. Please check the log for details.";
```
Comment (sfan5) [inline src/client/client.cpp:1773]: users are not the people who write the shaders
I suggest something like `"Failed to update textures! Shader compiling probably failed, check the log for details."`

File: src/client/shader.cpp [hunk 3]
```diff
@@ -462,7 +462,7 @@ u32 ShaderSource::getShaderIdDirect(const std::string &name,
 	infostream<<"getShaderIdDirect(): "
 			<<"Returning id="<<id<<" for name \""<<name<<"\""<<std::endl;
 
-	return id;
+	return info.shader_failed ? 0 : id;
```
Comment (sfan5) [inline src/client/shader.cpp:465]: are you sure 0 can't be a valid shader id?

Comment (lisacvuk) [reply src/client/shader.cpp:465]: The rest of the function seems to return 0 on failure https://github.com/minetest/minetest/blob/fc083643cb7f4a6697c6a45d70c0dbe6324b555c/src/client/shader.cpp#L427

File: src/client/client.cpp (hunk 2 above)
Comment (lisacvuk) [reply src/client/client.cpp]: Updated

**Comment by Wuzzy2** [comment]
@lisacvuk: Rebase needed.

**Comment by lisacvuk** [comment]
@Wuzzy2 Rebased, hopefully properly :/

**Comment by Wuzzy2** [comment]
@SmallJoker and other devs: In the future, please write a comment when adding the “Rebase needed” label, because as far I know, only comments trigger notifications, whereas label changes do not. If you only add the label but without comment, the PR author won't be notified. (If you're in a hurry, the comment can literally just be “Rebase needed.”, the point is just to trigger a notification.)

File: src/client/shader.cpp [hunk 4]
```diff
@@ -469,7 +469,7 @@ u32 ShaderSource::getShaderIdDirect(const std::string &name,
...
```
Comment (appgurueu) [inline src/client/shader.cpp:472]: Suggestion: Add the error message here. `if (info.shader_failed) client->setFatalError("Failed to compile " + name + "! Check the log for details.");`

Comment (lisacvuk) [reply src/client/shader.cpp:472]: I would have to pass the client pointer to ShaderSource, which seems like it would be a mess :/

**Comment by SmallJoker** [comment]
@Wuzzy2 Acknowledged.

Exit to main menu triggers a crash. To reproduce, enable shaders and add a typo in `nodes_shader/opengl_vertex.glsl`:
```
double free or corruption (!prev)
--Type <RET> for more, q to quit, c to continue without paging--

Thread 1 "minetest" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: Datei oder Verzeichnis nicht gefunden.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff6e0f859 in __GI_abort () at abort.c:79
#2  0x00007ffff6e7a3ee in __libc_message (action=action@entry=do_abort, 
    fmt=fmt@entry=0x7ffff6fa4285 "%s\n") at ../sysdeps/posix/libc_fatal.c:155
#3  0x00007ffff6e8247c in malloc_printerr (
    str=str@entry=0x7ffff6fa6690 "double free or corruption (!prev)") at malloc.c:5347
#4  0x00007ffff6e8412c in _int_free (av=0x7ffff6fd5b80 <main_arena>, p=0x55555c706020, 
    have_lock=<optimized out>) at malloc.c:4317
#5  0x00007ffff7cb4776 in irr::scene::CSceneManager::removeAll() ()
   from /data/Minetest/irrlicht/lib/Linux/libIrrlicht.so.1.9
#6  0x0000555555676eaa in ClientLauncher::run(GameStartData&, Settings const&) ()
#7  0x0000555555649fb0 in main ()
```

**Comment by lisacvuk** [comment]
> Exit to main menu triggers a crash. To reproduce, enable shaders and add a typo in `nodes_shader/opengl_vertex.glsl`
I'm getting this same crash on current master branch - add a typo to nodes_shader, join into a world, quit to main menu - seems to happen most attempts, but not every attempt. Can anyone confirm?

**Review by Df458 — CHANGES_REQUESTED** [review]
Ran some quick tests. I wasn't able to repro @SmallJoker's crash with the affected shaders, these all exited back to the menu properly. However, some new shaders have been added since and
- These aren't handled properly, most likely because this change has a separate check in each area we get a shader source
- When exiting to the menu after seeing this issue, I *do* get a crash. 

Unfortunately, it seems that due to its age this PR has had the goalposts moved on it... with that said, if you're still interested I'd like to see this updated to work with the new shadow shaders. Additionally, I wonder if maybe this is a candidate for refactoring--I'm a little uneasy about the individual checks scattered around everywhere, I suspect the next time someone wants to use a shader they may not know to add such a check.

**Pull request closed without merging**